python/dgl/distributed/sampling.py

```diff
@@ -46,52 +68,113 @@ def __getstate__(self):
+    the sampled edges are stored as the `dgl.EID` feature in the returned graph.
+
+    Parameters
+    ----------
+    g : DistGraph
+        The distributed graph.
+    nodes : tensor or dict
+        Node ids to sample neighbors from. The allowed types
+        are dictionary of node types to node id tensors, or simply node id tensor if
+        the given graph g has only one type of nodes.
+    fanout : int or dict[etype, int]
+        The number of sampled neighbors for each node on each edge type. Provide a dict
+        to specify different fanout values for each edge type.
+    edge_dir : str, optional
+        Edge direction ('in' or 'out'). If is 'in', sample from in edges. Otherwise,
+        sample from out edges.
+    prob : str, optional
+        Feature name used as the probabilities associated with each neighbor of a node.
+        Its shape should be compatible with a scalar edge feature tensor.
+    replace : bool, optional
+        If True, sample with replacement.
+
+    Returns
+    -------
+    DGLHeteroGraph
+        A sampled subgraph containing only the sampled neighbor edges from
+        ``nodes``. The sampled subgraph has the same metagraph as the original
+        one.
+    """
     assert edge_dir == 'in'
     req_list = []
     partition_book = dist_graph.get_partition_book()
-    np_nodes = toindex(nodes).tousertensor()
-    partition_id = partition_book.nid2partid(np_nodes)
+    nodes = toindex(nodes).tousertensor()
+    partition_id = partition_book.nid2partid(nodes)
+    LocalSampledGraph = namedtuple('LocalSampledGraph', 'global_src global_dst global_eids')
+    local_nids = None
     for pid in range(partition_book.num_partitions()):
-        node_id = F.boolean_mask(np_nodes, partition_id == pid)
-        if len(node_id) != 0:
+        node_id = F.boolean_mask(nodes, partition_id == pid)
+        # We optimize the sampling on a local partition if the server and the client
+        # run on the same machine. With a good partitioning, most of the seed nodes
+        # should reside in the local partition. If the server and the client
+        # are not co-located, the client doesn't have a local partition.
+        if pid == partition_book.partid and dist_graph.local_partition is not None:
+            assert local_nids is None
+            local_nids = node_id
+        elif len(node_id) != 0:
             req = SamplingRequest(node_id, fanout, edge_dir=edge_dir,
                                   prob=prob, replace=replace)
             req_list.append((pid, req))
-    res_list = remote_call_to_machine(req_list)
+
+    # send requests to the remote machine.
+    num_res = 0
+    msgseq2pos = None
+    if len(req_list) > 0:
+        num_res, msgseq2pos = send_requests_to_machine(req_list)
+
+    # sample neighbors for the nodes in the local partition.
+    res_list = []
+    if local_nids is not None:
+        src, dst, eids = _sample_neighbors(dist_graph.local_partition, partition_book,
```

Comment:
Local sample can only work with single process?

Reply:
dgl.distributed.sample_neighbors will be called in multiple processes.